anxcloud/resource_dns_record.go

```diff
@@ -0,0 +1,239 @@
+package anxcloud
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"go.anx.io/go-anxcloud/pkg/api"
+	"go.anx.io/go-anxcloud/pkg/api/types"
+	clouddnsv1 "go.anx.io/go-anxcloud/pkg/apis/clouddns/v1"
+	"go.anx.io/go-anxcloud/pkg/utils/object/compare"
+)
+
+var syncDNSRecordOps sync.Mutex
+var dnsRecordSkipMutexLock = providerContextKey("dns-record-skip-mutex-lock")
+
+func resourceDNSRecord() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceDNSRecordCreate,
+		ReadContext:   resourceDNSRecordRead,
+		UpdateContext: resourceDNSRecordUpdate,
+		DeleteContext: resourceDNSRecordDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(30 * time.Minute),
+			Read:   schema.DefaultTimeout(1 * time.Minute),
+			Update: schema.DefaultTimeout(10 * time.Minute),
+			Delete: schema.DefaultTimeout(5 * time.Minute),
+		},
+		Schema: schemaDNSRecord(),
+	}
+}
+
+func resourceDNSRecordCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	if v := ctx.Value(dnsRecordSkipMutexLock); v == nil {
+		syncDNSRecordOps.Lock()
+		defer syncDNSRecordOps.Unlock()
+	}
+	time.Sleep(time.Second)
+
+	a := m.(providerContext).api
+
+	r := dnsRecordFromResourceData(d)
+
+	// try to import
+	ret, err := findDNSRecord(ctx, a, r)
+	if err != nil {
```

Comment:
do you know `api.IgnoreNotFound()`? It's a bit more readable even in this case already :)

```go
if api.IgnoreNotFound(err) != nil {
    // error handling
} else if err != nil {
    // we have an error that wasn't ignored
    // no matching record found -> create new
} else {
    // record found -> import
}
```

Reply:
that's useful in many places of your code I guess :)

Reply:
I changed it to use the `api.IgnoreNotFound` helper but I think it's less readable now.
What about something like:
```go
if IsNotFound(err) {
	// create new
} else if err != nil {
	// handle error
} else {
	// import
}
```

Reply:
I found some suboptimal usages in the review I just submitted, maybe that handles that already - if not, we'll get back here :)

Reply:
do you still think it's less readable or can we resolve this?